### PR TITLE
add MarshalJSON and UnmarshalJSON to BitArray

### PR DESCRIPTION
**Pull request by melekes**
Refs https://github.com/tendermint/tendermint/issues/795

### CODE REVIEW

**Comment by jaekwon** [comment]
I don't think we should do this with MarshalJSON.
For developers, the existing struct's natural Amino:JSON format is fine.
Turning it into a custom formatted string makes it more difficult to create a compatible BitArray implementation in other languages.

File: common/bit_array.go [hunk 1]
```diff
@@ -320,3 +322,58 @@ func (bA *BitArray) Update(o *BitArray) {
...
+// JSON description.
+func (bA *BitArray) UnmarshalJSON(b []byte) error {
+	if string(b) == "null" {
+		bA = nil
+		return nil
+	}
+
+	bAStr := string(b)
+
+	matches := bitArrayJSONRegexp.FindAllStringSubmatch(bAStr, 2)
```
Comment (jaekwon) [inline common/bit_array.go]: There's no need to use All here, cuz the regexp is \A \z

File: common/bit_array.go [hunk 2]
```diff
@@ -320,3 +322,58 @@ func (bA *BitArray) Update(o *BitArray) {
...
+	return []byte("\"" + bA.stringIndented("") + "\""), nil
+}
+
+var bitArrayJSONRegexp = regexp.MustCompile(`\A"BA{(\d+):([_X\s]+)}"\z`)
+
+// UnmarshalJSON implements json.Unmarshaler interface by unmarshaling a custom
+// JSON description.
+func (bA *BitArray) UnmarshalJSON(b []byte) error {
+	if string(b) == "null" {
+		bA = nil
```
Comment (jaekwon) [inline common/bit_array.go]: This doesn't do anything.

**Comment by ebuchman** [comment]
💃

**Pull request merged**